ext/flash/lib.rs

```
@@ -0,0 +1,1451 @@
+    if nread == 0 {
+      let _ = sock.write(b"0\r\n\r\n");
+      break;
+    }
+    let response = &buf[..nread];
+
+    let _ = sock.write(format!("{:x}", response.len()).as_bytes());
+    let _ = sock.write(b"\r\n");
+    let _ = sock.write(response);
+    let _ = sock.write(b"\r\n");
+  }
+
+  resource.close();
+  Ok(())
+}
+
+pub struct RespondChunkedFast;
+
+impl fast_api::FastFunction for RespondChunkedFast {
+  fn function(&self) -> *const c_void {
+    op_flash_respond_chunked_fast as *const c_void
+  }
+
+  fn args(&self) -> &'static [fast_api::Type] {
+    &[
+      fast_api::Type::V8Value,
+      fast_api::Type::Uint32,
+      fast_api::Type::TypedArray(fast_api::CType::Uint8),
+      fast_api::Type::Bool,
+    ]
+  }
+
+  fn return_type(&self) -> fast_api::CType {
+    fast_api::CType::Void
+  }
+}
+
+unsafe fn op_flash_respond_chunked_fast(
+  recv: v8::Local<v8::Object>,
+  token: u32,
+  response: *const fast_api::FastApiTypedArray<u8>,
+  shutdown: bool,
+) {
+  let ptr =
+    recv.get_aligned_pointer_from_internal_field(V8_WRAPPER_OBJECT_INDEX);
+  let ctx = &mut *(ptr as *mut ServerContext);
+
+  let response = &*response;
+  if let Some(response) = response.get_storage_if_aligned() {
+    respond_chunked(ctx, token, shutdown, Some(response));
+  } else {
+    todo!();
+  }
+}
+
+fn respond_chunked(
+  ctx: &mut ServerContext,
+  token: u32,
+  shutdown: bool,
+  response: Option<&[u8]>,
+) {
+  let sock = match shutdown {
+    true => {
+      let tx = ctx.response.remove(&token).unwrap();
+      tx.socket()
+    }
+    // In case of a websocket upgrade or streaming response.
+    false => {
+      let tx = ctx.response.get(&token).unwrap();
+      tx.socket()
+    }
+  };
+
+  if let Some(response) = response {
+    let _ = sock.write(format!("{:x}", response.len()).as_bytes());
+    let _ = sock.write(b"\r\n");
+    let _ = sock.write(response);
+    let _ = sock.write(b"\r\n");
```

multiple syscalls?

---

Ideally we’d use write_all_vectored() here.
Also I think we shouldn’t ignore errors.